text/2018-11-12-backup-restore-replication-with-raft-learner.md

```diff
@@ -0,0 +1,168 @@
+
+**In the following example, we will still use Leader for snapshot and replication, and use Ceph to represent the remote storage.**
+
+When a new learner of Region 1 is added to the Backup node, it will:
+
+1. Create a Raft peer and save the Region meta in the local RocksDB.
+2. Receive the snapshot from the leader and save it to the Ceph.
+3. Persist the Raft state (like applied index, committed Index) in the local RocksDB. Of course, we must also record how to find the snapshot file in Ceph :-)
+4. Receive the Raft logs from the leader, and append them to the Ceph. For simplicity, we can use one file for one Region and append the logs to it. To ensure higher safety and to avoid our foolish mistakes like writing wrong data to the state machine, we need to keep Raft logs as more as possible, maybe keep all logs and don’t compact them. 
+5. Apply the committed logs. Unlike the common way which writes all data to the state machine, here we just ignore most of the commands, and only care some Admin commands like Split, Merge and ConfChange.
+    + The major reason that we don’t want to apply data to the state machine is performance. Unlike sequential write of Raft logs, these writes maybe random and definitely will cause a high pressure for the storage.
+    + For Split, we also need to create the new split Region
+    + For Merge, we also need to destroy the merged Region.
+    + For CompactLog, we can ignore this command directly.
+    + For ConfChange removing the learner in the Backup node, we also need to destroy the Region. 
+6. Every day (or other regular times), the learner can request a new snapshot from the leader, which can speed up later Backup and Restore. 
+
+
+All look fine, but there is a problem for step 5. We need to handle some corner cases for the Admin commands like Split. I will discuss it in the Restore section later. 
+
+For better performance, we can cache the Raft logs locally and then write them to Ceph in batches.
+
+## Backup
+
+If we want to do the backup, we can send a Checkpoint command to the Backup node, and the node will do:
+
+1. Let all the learner sync the latest committed logs from leaders. This can be done by Follower read mechanism.
+2. Get a snapshot of the current node. Because the Backup node uses RocksDB, we can easily support it. In the snapshot, we can know current all learners’ metadata, the associated snapshot files, the Raft logs, etc.
+
+For the Backup, we only need to retrieve the metadata. It is lightweight, so we can do the backup frequently. 
+
+## Restore
+
+After Backup, we can get a snapshot of the whole cluster, we can use it for restore. 
+
+For a Region, if we get a snapshot file and its following Raft logs, we can first apply the snapshot, and then apply the Raft logs in order, and the Region will have the consistent data eventually. We name this flow Replay later.
```

Comment:
Since we're giving "Replay" a name maybe you can bold it or emphasize it in some way, so readers know it is an important term.